### PR TITLE
feat(sweeps): optuna supports multi-objective optimization

### PR DESCRIPTION
**Pull request by gtarpenning**
https://wandb.atlassian.net/browse/WB-14249

### CODE REVIEW

File: jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py [hunk 1]
```diff
@@ -92,18 +98,20 @@ def _handle_job_logic(run, name, enable_git=False) -> None:
         )
         tag = os.environ.get("WANDB_DOCKER", "").split(":")
         if len(tag) == 2:
-            jobstr += f"-{tag[0].replace('/', '_')}_{tag[-1]}:latest"
-        else:
+            jobstr += f"-{tag[0].replace('/', '_')}:{tag[-1]}"
+        elif len(tag) == 1:
+            jobstr += f"-{tag[0].replace('/', '_')}:latest"
+        else:  # unknown format
```
Comment (gtarpenning) [inline jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py:104]: update for new image job tag format

File: jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py [hunk 2]
```diff
@@ -80,34 +89,6 @@ def setup_scheduler(scheduler: Scheduler, **kwargs):
...
-        wandb.termlog(f"Creating image job {click.style(jobstr, fg='yellow')}\n")
-    elif not enable_git:
-        jobstr += f"-{name}:latest"
-        wandb.termlog(
-            f"Creating code-artifact job: {click.style(jobstr, fg='yellow')}\n"
-        )
-    else:
-        _s = click.style(f"https://wandb.ai/{jobstr}s/", fg="yellow")
-        wandb.termlog(f"Creating repo-artifact job found here: {_s}\n")
-        run.log_code(name=name, exclude_fn=lambda x: x.startswith("_"))
```
Comment (gtarpenning) [inline jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py:107]: now handled by bens pr

File: jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py [hunk 3]
```diff
@@ -62,26 +62,26 @@ def setup_scheduler(scheduler: Scheduler, **kwargs):
...
         project=cli_args.project,
         entity=cli_args.entity,
     )
     config = run.config
+    args = config.get("sweep_args", {})
 
-    if not config.get("sweep_args", {}).get("sweep_id"):
-        # not a sweep, just finish the run and return
+    if not args or not args.get("sweep_id"):
+        # when the config has no sweep args, this is being run directly from CLI
```
Comment (KyleGoyette) [inline jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py:77]: So the user is running the `wandb launch -j <job>` here? Is this specifically done to create the job?

File: jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py [hunk 4]
```diff
@@ -175,14 +178,12 @@ def formatted_trials(self) -> str:
...
-                    continue
-
                 if len(trial.values) != len(self._metric_defs):
                     wandb.termwarn(
-                        f"{LOG_PREFIX}Number of trial metrics ({trial.values})"
+                        f"{LOG_PREFIX}Number of logged metrics ({trial.values})"
                         " does not match number of metrics defined "
-                        f"({self._metric_defs})"
+                        f"({self._metric_defs}). Specify metrics for optimization"
+                        " in the scheduler.settings.metrics portion of the sweep config"
```
Comment (KyleGoyette) [inline jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py:186]: What would a user have done to create this case?

**Review by KyleGoyette — COMMENTED** [review]
A couple more questions

File: jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py (hunk 3 above)
Comment (gtarpenning) [reply jobs/sweep_schedulers/optuna_scheduler/optuna_scheduler.py:77]: now this would be triggered by running `python optuna_scheduler.py`

**Review by KyleGoyette — APPROVED** [review]
🔥

**Pull request merged**